File: Difficulty/Medium/[46] Permutations/solution.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        size = len(nums)
        self.permutation(res, 0, size, nums)
        return res
    
    def permutation(self, res, begin, size, nums):
        if begin == size - 1:
            res.append(nums[:])
            return
        for i in range(begin, size):
            nums[begin], nums[i] = nums[i], nums[begin]
            self.permutation(res, begin + 1, size, nums)
            nums[begin], nums[i] = nums[i], nums[begin]

File: Difficulty/Medium/[46] Permutations/solution.cpp
/*
 * @lc app=leetcode id=46 lang=cpp
 *
 * [46] Permutations
 */
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        int size = nums.size();
        permutation(res, 0, size - 1, nums);
        return res;
    }
    
    void permutation(vector<vector<int>>& res, int begin, int len, vector<int>& nums){
        if (begin == len){
            res.push_back(nums);
            return;
        }
        
        for (int i = begin; i <= len; ++i){
            swap(nums[begin], nums[i]);
            permutation(res, begin + 1, len, nums);
            swap(nums[begin], nums[i]);
        }
    }
};

